pinot-common/src/main/java/org/apache/pinot/common/utils/http/HttpClient.java

```diff
@@ -394,6 +395,29 @@ public int downloadFile(URI uri, int socketTimeoutMs, File dest, AuthProvider au
     }
   }
 
+  public File downloadUntarFileStreamed(URI uri, int socketTimeoutMs, File dest, AuthProvider authProvider,
+      List<Header> httpHeaders, long rateLimit)
+      throws IOException, HttpErrorStatusException {
+    HttpUriRequest request = getDownloadFileRequest(uri, socketTimeoutMs, authProvider, httpHeaders);
+    File ret;
+    try (CloseableHttpResponse response = _httpClient.execute(request)) {
+      StatusLine statusLine = response.getStatusLine();
+      int statusCode = statusLine.getStatusCode();
+      if (statusCode >= 300) {
+        throw new HttpErrorStatusException(HttpClient.getErrorMessage(request, response), statusCode);
+      }
+
+      try (InputStream inputStream = response.getEntity().getContent()) {
+        ret = TarGzCompressionUtils.untarRateLimit(inputStream, dest, rateLimit).get(0);
+      }
+
```

Comment:
Is there a way to do a sanity check / validation when streaming untar finishes to ensure the download was correct / complete ?

Reply:
Discussed offline. The untar library will fail on checksum and report IOException if files are corrupted during transmission